test/ecto/changeset_test.exs

```diff
@@ -1358,6 +1358,18 @@ defmodule Ecto.ChangesetTest do
     end
   end
 
+  test "validate_number/3 with bad target" do
+    # Number value
+    assert_raise ArgumentError, "expected target value to be of type Integer or Float, got: #Decimal<10>", fn ->
+      validate_number(changeset(%{"upvotes" => 11}), :upvotes, greater_than: Decimal.new(10))
+    end
```

Comment:
I think this should be valid. If we allow decimal to int, why not int to decimal for consistency? Code would also be simpler.

Reply:
I think I understand what you mean. Could you let me know if this makes sense:

There are several ways the types of `value` and `target_value` can be combined. They can both be `number`, `Decimal`, or something else. So 9 validation scenarios.

a) Both are decimal (1 scenario) - should do a decimal compare
b) Both are number (1 scenario) - should do a `spec_function` compare
c) One is decimal and one is number (2 scenarios) - should do a decimal compare
d) One is decimal and one is not a number/Decimal (2 scenarios) - should raise
e) One is number and one is not a number/Decimal (2 scenarios) - should raise
f) Both are not number/Decimal (1 scenario) - should raise

Reply:
I think you can decouple the validation from the operation. The validation is that any can be a number or a decimal. Then the operation is, if any is a decimal, decimal op, otherwise Kernel op. Does this make sense?

Reply:
Makes sense! I will re-submit shortly. Thanks both.

Reply:
I made the changes but was wondering what you guys think of the style. I'm not if you prefer more pattern matches and simpler guards or complex guards and fewer pattern matches. The complex guards look a bit awkward to me but happy change.

Reply:
I think we might be able to clean it up but I am not sure. I will play with it a bit locally. :)